src/driver.rs

```diff
@@ -120,110 +122,145 @@ pub fn block_on<T>(future: impl Future<Output = T>) -> T {
         unparker().unpark();
     });
 
-    // Parker and unparker for notifying the current thread.
-    let (p, u) = parking::pair();
-    // This boolean is set to `true` when the current thread is blocked on I/O.
-    let io_blocked = Arc::new(AtomicBool::new(false));
+    // Creates a parker and an associated waker that unparks it.
+    fn parker_and_waker() -> (Parker, Waker, Arc<AtomicBool>) {
+        // Parker and unparker for notifying the current thread.
+        let (p, u) = parking::pair();
+
+        // This boolean is set to `true` when the current thread is blocked on I/O.
+        let io_blocked = Arc::new(AtomicBool::new(false));
+
+        // Prepare the waker.
+        let waker = waker_fn({
+            let io_blocked = io_blocked.clone();
+            move || {
+                if u.unpark() {
+                    // Check if waking from another thread and if currently blocked on I/O.
+                    if !IO_POLLING.with(Cell::get) && io_blocked.load(Ordering::SeqCst) {
+                        Reactor::get().notify();
+                    }
+                }
+            }
+        });
+
+        (p, waker, io_blocked)
+    }
 
     thread_local! {
+        // Cached parker and waker for efficiency.
+        static CACHE: RefCell<(Parker, Waker, Arc<AtomicBool>)> = RefCell::new(parker_and_waker());
+
         // Indicates that the current thread is polling I/O, but not necessarily blocked on it.
         static IO_POLLING: Cell<bool> = Cell::new(false);
     }
 
-    // Prepare the waker.
-    let waker = waker_fn({
-        let io_blocked = io_blocked.clone();
-        move || {
-            if u.unpark() {
-                // Check if waking from another thread and if currently blocked on I/O.
-                if !IO_POLLING.with(Cell::get) && io_blocked.load(Ordering::SeqCst) {
-                    Reactor::get().notify();
-                }
+    return CACHE.with(|cache| {
+        // Try grabbing the cached parker and waker.
+        match cache.try_borrow_mut() {
+            Ok(cache) => {
+                // Use the cached parker and waker.
+                let (parker, waker, io_blocked) = &*cache;
+                block_on_inner(future, parker, waker, io_blocked)
+            }
+            Err(_) => {
+                // Looks like this is a recursive `block_on()` call.
+                // Create a fresh parker and waker.
```

Comment:
Could you add a test making sure recursive `block_on` calls work? Make sure to test cases where the outer `block_on` is notified first and vice versa.